Encoder/DefaultEncoder.php

```diff
@@ -0,0 +1,163 @@
+<?php
+
+namespace Lexik\Bundle\JWTAuthenticationBundle\Encoder;
+
+use InvalidArgumentException;
+use Lexik\Bundle\JWTAuthenticationBundle\Services\KeyLoader\KeyLoaderInterface;
+use Namshi\JOSE\JWS;
+
+/**
+ * Default class for encode/decode JWT, based on the Namshi/JOSE library.
+ * Supports OpenSSL and PHPSecLib encryption engines.
+ *
+ * @since 2.0
+ * @author Robin Chalas <robin.chalas@gmail.com>
+ */
+class DefaultEncoder extends JWTEncoder
+{
+    /**
+     * @var KeyLoaderInterface
+     */
+    protected $keyLoader;
+
+    /**
+     * @var string
+     */
+    protected $encryptionEngine;
+
+    /**
+     * @var string
+     */
+    protected $encryptionAlgorithm;
+
+    /**
+     * Constructor.
+     *
+     * @param KeyLoaderInterface $keyLoader
+     * @param string             $encryptionEngine
+     * @param string             $encryptionAlgorithm
+     *
+     * @throws InvalidArgumentException If the given encryptionEngine is not supported.
+     * @throws InvalidArgumentException If the given algorithm is not supported.
+     */
+    public function __construct(KeyLoaderInterface $keyLoader, $encryptionEngine, $encryptionAlgorithm)
+    {
+        if (!in_array($encryptionEngine, ['OpenSSL', 'SecLib'])) {
+            throw new InvalidArgumentException(
+                sprintf('The encryption engine "%s" is not supported', $encryptionEngine)
+            );
+        }
+
+        if (!$this->isAlgorithmSupportedForEngine($encryptionEngine, $encryptionAlgorithm)) {
+            throw new InvalidArgumentException(
+                sprintf('The algorithm "%s" is not supported for %s', $encryptionAlgorithm, $encryptionEngine)
+            );
+        }
+
+        $this->keyLoader           = $keyLoader;
+        $this->encryptionEngine    = $encryptionEngine;
+        $this->encryptionAlgorithm = $encryptionAlgorithm;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws \RuntimeException If the JWS cannot be signed
+     */
+    public function encode(array $data)
+    {
+        $jws = new JWS(['alg' => $this->encryptionAlgorithm], $this->encryptionEngine);
```

Comment:
What do you think about actually configuring the `JWS` instance as a provider service and then injecting that directly into this class? If we added an additional configuration param - `encoder_algorithm` - that defaults to `OpenSSL`, we would only need to register *one* `JWS` service and *one* encoder service (instead of 2). Is there any reason why someone would want to have two services available at once (`OpenSSL` AND `SecLib`)? If not, then one service would be better.